### PR TITLE
Feature/symbol table

### PR DESCRIPTION
**Pull request by markwpearce**
Adds the `SymbolTable` class, and builds the symbol table from assignments and function names

### CODE REVIEW

File: src/SymbolTable.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import type { Identifier } from './lexer/Token';
+import type { BscType } from './types/BscType';
+import { DynamicType } from './types/DynamicType';
+import { UninitializedType } from './types/UninitializedType';
+
+export interface BscSymbol {
```
Comment (TwitchBronBron) [inline src/SymbolTable.ts:6]: Can you move this interface to the bottom so the primary export (i.e. SymbolTable) is at the top of the file?

File: src/SymbolTable.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+     * @returns {BscSymbol}
+     */
+    getSymbol(name: string): BscSymbol[] {
+        const key = name.toLowerCase();
+        return this.symbols.get(key) ?? this.parent?.getSymbol(key);
+    }
+
+    /**
+     * Adds a new symbol to the table
+     * @param {Identifier} name
```
Comment (TwitchBronBron) [inline src/SymbolTable.ts:46]: You don't need to specify the data type (i.e. `{Identifier}`) with jsdocs in typescript...any docs generator would infer the type from the typed parameters below.

File: src/SymbolTable.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    /**
+     * Adds all the symbols from another table to this one
+     * It will overwrite any existing symbols in this table
+     * @param symbolTable
+     */
+    mergeSymbolTable(symbolTable: SymbolTable) {
+        for (let [key, value] of symbolTable.symbols) {
+            this.symbols.set(key, value);
```
Comment (TwitchBronBron) [inline src/SymbolTable.ts:94]: We should merge the symbols, not overwrite them. Call `this.addSymbol` instead.

File: src/parser/Parser.spec.ts [hunk 4]
```diff
@@ -1124,6 +1127,71 @@ describe('parser', () => {
...
+                return "foo"
+                end function
+            `, ParseMode.BrighterScript);
+            expectZeroDiagnostics(parser.diagnostics);
+            const someFuncSymbolTable = parser.references.functionExpressions[0].symbolTable;
+            expect(someFuncSymbolTable.getSymbolType('temp')).to.be.instanceof(LazyType);
+            expect(someFuncSymbolTable.getSymbolType('temp').toTypeString()).to.eq('string');
+        });
+
+        it('does not know about symbols declared in parent functions', () => {
```
Comment (TwitchBronBron) [inline src/parser/Parser.spec.ts:1181]: Nice test!

File: src/parser/Parser.spec.ts [hunk 5]
```diff
@@ -1124,6 +1127,71 @@ describe('parser', () => {
             expect(type.returnType).to.be.instanceof(CustomType);
         });
     });
+
+    describe('symbolTable', () => {
```
Comment (TwitchBronBron) [inline src/parser/Parser.spec.ts:1131]: Can you add tests to ensure `for` and `for-each` loop variable are included in the symbol table? 

`for i = 0 to 10 step 1` and `for each thing in items`

**Review by TwitchBronBron — CHANGES_REQUESTED** [review]
Looks great! I have items that need to be addressed.

File: src/SymbolTable.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        for (let [, value] of symbolTable.symbols) {
+            for (const symbol of value) {
+                this.addSymbol(symbol.name, symbol.type);
+            }
+        }
+    }
+}
+
+
+export interface BscSymbol {
```
Comment (TwitchBronBron) [inline src/SymbolTable.ts:98]: Let's change this interface to:
```
export interface BscSymbol {
    name: string;
    range: Range;
    type: BscType;
}
```

That way, we can add any string-based symbol as long as we can generate a range for it, such as namespaces which would have multiple identifiers, or other future symbols we haven't realized yet. 

```
          |identifier | dot | identifier |
namespace    Animals     .       Dogs
```

File: src/SymbolTable.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        const key = name.toLowerCase();
+        return this.symbols.get(key) ?? this.parent?.getSymbol(key);
+    }
+
+    /**
+     * Adds a new symbol to the table
+     * @param name
+     * @param  type
+     */
+    addSymbol(name: Identifier, type: BscType) {
```
Comment (TwitchBronBron) [inline src/SymbolTable.ts:45]: Based on [this comment](https://github.com/rokucommunity/brighterscript/pull/380/files#r605556107), change this method signature to:
```
addSymbol(name: string, type: BscType, range?: Range) {
```

**Comment by markwpearce** [comment]
Not sure why the mac build failed? I assume some sort of external issue

**Comment by TwitchBronBron** [comment]
> Not sure why the mac build failed? I assume some sort of external issue - I have a Mac, and it builds/lints/runs fine on my machine

The GitHub actions mac builds sometimes have issues. I re-ran them and everything is passing now.

**Comment by TwitchBronBron** [comment]
This is a great step forward! Per our conversations in slack, we still have a few scope-related changes to make before the target PR can be merged. However, I'm happy with the current progress of this PR so I'm merging it.

**Pull request merged**